tools/extra/fpgadiag/common.py

```diff
@@ -221,19 +261,87 @@ def get_port_list(self, argport, total):
+        with open(pci_dev_path, "r+b", 0) as f:
+            mm = mmap.mmap(f.fileno(), MAPSIZE, mmap.MAP_SHARED,
+                           mmap.PROT_WRITE, 0, base)
+            # write data (64 bit)
+            mm[offset:offset+8] = data
+            # close mapping
+            mm.close()
+
+    def pci_bar2_rw(self, addr, data=None):
+        pci_bar2_path = '/sys/bus/pci/devices/{}/resource2'
+        if data is None:    # read
+            return self.pci_read(pci_bar2_path.format(self.sbdf), addr)
+        else:   # write
+            self.pci_write(pci_bar2_path.format(self.sbdf), addr, data)
+
+    def upl_indirect_rw(self, addr, data=None):
+        addr &= 0xffffff
+        rdata = 0
+        if data is None:      # read
+            cmd = (1 << 62) | (addr << 32)
+            self.pci_bar2_rw(self.upl_base + UPL_INDIRECT_CTRL_REG, cmd)
+            while (rdata >> 32) != 0x1:     # waiting for read valid
+                rdata = self.pci_bar2_rw(self.upl_base + UPL_INDIRECT_DATA_REG)
+            return rdata & 0xffffffff
+        else:       # write
+            cmd = (2 << 62) | (addr << 32) | (data & 0xffffffff)
+            self.pci_bar2_rw(self.upl_base + UPL_INDIRECT_CTRL_REG, cmd)
+            if (rdata >> 32) != 0x1:     # waiting for write complete
+                rdata = self.pci_bar2_rw(self.upl_base + UPL_INDIRECT_DATA_REG)
+
+    def get_upl_base(self):
+        addr = 0
+        while True:
+            header = self.pci_bar2_rw(addr)
+            feature_type = (header >> DFH_TYPE_SHIFT) & DFH_TYPE_MASK
+            feature_id = (header >> DFH_ID_SHIFT) & DFH_ID_MASK
+            if feature_type == DFH_TYPE_AFU and feature_id == DFH_ID_UPL:
+                self.upl_base = addr
+                break
+            if feature_type in [DFH_TYPE_AFU, DFH_TYPE_FIU]:
+                next_afu_offset = self.pci_bar2_rw(addr+NEXT_AFU_OFFSET_REG)
+                next_afu_offset &= NEXT_AFU_OFFSET_MASK
+            if next_afu_offset == 0 or (next_afu_offset & 0xffff) == 0xffff:
+                print("Use default UPL base address {:#x}".format(
+                                                                self.upl_base))
+                break
+            else:
+                addr += next_afu_offset
+                next_afu_offset = 0
+
 
 def main():
     parser = argparse.ArgumentParser()
-    parser.add_argument('--segment', '-S', type=hexint,
-                        help='Segment number of PCIe device')
-    parser.add_argument('--bus', '-B', type=hexint,
-                        help='Bus number of PCIe device')
-    parser.add_argument('--device', '-D', type=hexint,
-                        help='Device number of PCIe device')
-    parser.add_argument('--function', '-F', type=hexint,
-                        help='Function number of PCIe device')
+    parser.add_argument('--segment')
+    parser.add_argument('--bus')
+    parser.add_argument('--device')
+    parser.add_argument('--function')
```

Comment:
help messages might be useful here. Not sure why it was taken out

Reply:
fixed